Session08/binarytree/solutions/BinarySearchTree.java

```diff
@@ -0,0 +1,123 @@
+// This class fails to compile, can you figure out why, and fix it?
+// This is beacuse BinaryTree has a constructor which is not default. Hence, when we extend this class and try to call a no-argument constructor, it'll give a compile-time error. so, we call it by passing the parameters (to instantiate the root node) in super() called inside the constructor of BinarySearchTree
+
+class BinarySearchTree extends BinaryTree {
+
+    private int flag = 0;
+    
+    public BinarySearchTree()
+    {
+        super(0);
+    }
+    
+    // insert node
+	@Override
+	public boolean insert(int value) {
+		super.root = insert(super.root, value);
+		return true;
+	}
+
+    private TreeNode insert(TreeNode root, int data)
+    {
+        // if current node is null, insert node with data
+        if (root == null) root = new TreeNode(data);
+        else
+        {
+            // traverse to left sub-tree if value to be inserted is lesser than value in current node
+            if (data <= node.getData()) root.left = insert(root.left, data);
+            // traverse to right sub-tree if value to be inserted is greater than value in current node
+            else root.right = insert(root.right, data);
+        }
+        return root;
+    }
+    
+    // delete node
```

Comment:
Is this comment really necessary? The function name is `delete`. Self-documenting code.